TTCN3_Project/CANFilter/com/testingtech/playits/canfilter/CANFilterService.java

```diff
@@ -0,0 +1,216 @@
+              + " request. " + e.getMessage());
+        }
+      }
+
+    } catch (IOException e) {
+      System.err.println("[CANFilterService] Error listening on port "
+          + address.getPort() + ": " + e);
+    } catch (JSONException e) {
+      System.err
+          .println("[CANFilterService] Error while reading json objects from input stream");
+    } finally {
+      System.out
+          .println("[CANFilterService] No more requests. Closing sockets");
+      close(socket);
+      close(serverSocket);
+    }
+  }
+
+  private static void close(ServerSocket serverSocket) {
+    try {
+      if (serverSocket != null) {
+        serverSocket.close();
+      }
+    } catch (IOException e) {
+      System.err.println("[CANFilterService] Could not close socket: "
+          + e.getMessage());
+    }
+  }
+
+  private static void close(Socket socket) {
+    try {
+      if (socket != null) {
+        socket.close();
+      }
+    } catch (IOException e) {
+      System.err.println("[CANFilterService] Could not close socket: "
+          + e.getMessage());
+    }
+  }
+
+  private static ServerSocket createServerSocket(InetSocketAddress address)
+      throws IOException, SocketException {
+    ServerSocket serverSocket = new ServerSocket();
+    serverSocket.setReuseAddress(true);
+    serverSocket.bind(address);
+    return serverSocket;
+  }
+
+  private static void handleRequest(JSONObject request) throws IOException,
+      JSONException {
+    JSONArray data = request.getJSONArray("reqData");
+    switch (request.getString("reqType")) {
+      case "start":
+        addEntries(data);
+        break;
+      case "stop":
+        removeEntries(data);
+        break;
+      default:
+        System.err.println("[CANFilterService] Unsupported request");
+        break;
+    }
+  }
+
+  private static void addEntries(JSONArray jsonArray) throws JSONException {
+    for (int index = 0; index < jsonArray.length(); index++) {
+      JSONObject jsonObject = jsonArray.getJSONObject(index);
+      addEntry(jsonObject.getString("key"), jsonObject.getInt("interval"));
+    }
+  }
+
+  private static void addEntry(String key, int interval) {
+    System.out.println("[CANFilterService] Adding entry " + key
+        + " with interval " + interval);
+    car2xEntries.put(key, new Car2XEntry());
+    if (entryUpdater == null || entryUpdater.getState() != State.RUNNABLE) {
+      System.out.println("[CANFilterService] Starting entry updater");
+      entryUpdater = new Car2XEntryUpdater(car2xEntries);
+      entryUpdater.start();
+    }
+    timeoutResponder.addTimer(key, interval);
+  }
+
+  private static void removeEntries(JSONArray jsonArray) throws JSONException {
+    for (int index = 0; index < jsonArray.length(); index++) {
+      JSONObject jsonObject = jsonArray.getJSONObject(index);
+      removeEntry(jsonObject.getString("key"));
+    }
+  }
+
+  private static void removeEntry(String key) {
```

Comment:
The only reason for the statics here was, that the methods are usable in the main method. But in my opinion that is against different principles of object oriented programming.